### ISSUE TITLE
Internal Server Error after latest upgrade

### ISSUE THREAD

**Issue by phutch11** [yeti-platform/yeti#587]
### Description
After performing the usual
```
cd /opt/yeti/
git pull
./yeti.py syncdb
```
I encountered the missing mongo-engine message. 

Checked that part of the latest updates included a updated requirements list.

`pip install -r requirements.txt --user `

Started getting the following for a number of requirements:
```
Collecting html2text==2020.1.16 (from -r requirements.txt (line 29))
  Could not find a version that satisfies the requirement html2text==2020.1.16 (from -r requirements.txt (line 29)) (from versions: 2.35, 2.37, 2.38, 3.1, 3.2, 3.101, 3.200.0, 3.200.1, 3.200.2, 3.200.3, 2014.4.5, 2014.7.3, 2014.9.7, 2014.9.8, 2014.9.25, 2014.12.5, 2014.12.24, 2014.12.29, 2015.2.18, 2015.4.13, 2015.4.14, 2015.6.6, 2015.6.12, 2015.6.21, 2015.11.4, 2016.1.8, 2016.4.2, 2016.5.29, 2016.9.19, 2017.10.4, 2018.1.9, 2019.8.11)
No matching distribution found for html2text==2020.1.16 (from -r requirements.txt (line 29))
The directory '/home/yeti/.cache/pip/http' or its parent directory is not owned by the current user and the cache has been disabled. Please check the permissions and owner of that directory. If executing pip with sudo, you may want sudo's -H flag.
The directory '/home/yeti/.cache/pip' or its parent directory is not owned by the current user and caching wheels has been disabled. check the permissions and owner of that directory. If executing pip with sudo, you may want sudo's -H flag.
Collecting uwsgi
```
Switched to pip3 and the requirements completed.

Tested site and usual checks that data could be accessed, modified, etc. All working.

However, after a reboot, the server returns an Internal Server Error.

Currently have a snapshot of the server but need to refresh the snapshot after making changes - cannot reboot server.

### Environment

| Question         | Answer
|------------------|--------------------
| Git commit       | https://github.com/yeti-platform/yeti/commit/bc6d7ab7374f3b015c2f92ed8f5ac7e303f95474
| OS version       | Ubuntu 18.04
| Browser          | Edge/18.18363 & Chrome 81.0.4044.138

### Steps to Reproduce
1. cd /opt/yeti/
2. git pull
3. ./yeti.py syncdb
- Got `ImportError: No module named mongoengine`
4. pip3 install -r requirements --user (from /opt/yeti)
5. (Site still accessible)
6. Reboot
7. Browse to URI and get immediate Internal Server Error

#### Expected behavior
See contents of the server/observable/ front page

#### Actual behavior
Page immediately returns small "Internal Server Error" (Not the big 500 one you sometimes see during a timeout)

Someone reported they had this same error and fixed it with permissions on the /var/log/yeti activity file. I was not successful in replicating this. Permissions are same on file/folder prior and post reboot

The Yeti activity logs did not contain any information about errors.

#### [Journal]
```
May 14 17:02:07 ubuntu uwsgi[781]: --- no python application found, check your startup logs for errors ---
May 14 17:02:07 ubuntu uwsgi[781]: [pid: 1038|app: -1|req: -1/2] 192.168.10.111 () {42 vars in 691 bytes} [Thu May 14 17:02:07 2020] GET /observable/ => generated 21 bytes in 0 msecs (HTTP/1.1 500) 2 hea
May 14 17:03:26 ubuntu dbus-daemon[1005]: [session uid=1000 pid=1005] Activating via systemd: service name='org.gnome.Terminal' unit='gnome-terminal-server.service' requested by ':1.78' (uid=1000 pid=197
```
And not sure if they are of any help but also found these:
```
May 14 17:29:56 ubuntu systemd[1]: yeti_analytics.service: Main process exited, code=exited, status=1/FAILURE
May 14 17:29:56 ubuntu systemd[1]: yeti_analytics.service: Failed with result 'exit-code'.
May 14 17:29:56 ubuntu systemd[1]: yeti_oneshot.service: Main process exited, code=exited, status=1/FAILURE
May 14 17:29:56 ubuntu systemd[1]: yeti_oneshot.service: Failed with result 'exit-code'.
May 14 17:29:56 ubuntu systemd[1]: yeti_feeds.service: Main process exited, code=exited, status=1/FAILURE
May 14 17:29:56 ubuntu systemd[1]: yeti_feeds.service: Failed with result 'exit-code'.
May 14 17:29:56 ubuntu /usr/lib/gdm3/gdm-x-session[862]: dbus-update-activation-environment: setting QT_ACCESSIBILITY=1
May 14 17:29:56 ubuntu systemd[1]: yeti_beat.service: Main process exited, code=exited, status=1/FAILURE
May 14 17:29:56 ubuntu systemd[1]: yeti_beat.service: Failed with result 'exit-code'.
May 14 17:29:56 ubuntu systemd[1]: yeti_exports.service: Main process exited, code=exited, status=1/FAILURE
May 14 17:29:56 ubuntu systemd[1]: yeti_exports.service: Failed with result 'exit-code'.
```
#### [nginx/access_log]
192.168.10.111 - - [14/May/2020:16:58:49 +1000] "GET /observable/ HTTP/1.1" 500 32 "-" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/70.0.3538.102 Safari/537.36 Edge/18.18363"

**Comment by tomchop** [comment]
Hmm I think the main problem lies in `no python application found` in your uwsgi logs. Can you share the uwsgi configuration? We haven't done any changes that should affect this though.

Maybe the uwsgi daemon does not have access to the code files anymore?

**Comment by phutch11** [comment]
I've had a hunt around and I don't believe there's been any additional custom configurations made.

This server was originally built using the ubuntu_bootstrap method which worked great back then (not so much today since the upgrade).

The current settings for the `yeti_uwsgi.service` are:
```
[Unit]
Description=Yeti UWSGI server
After=mongodb.service redis.service

[Service]
Type=simple
User=yeti
WorkingDirectory=/opt/yeti
ExecStart=/usr/local/bin/uwsgi --socket 127.0.0.1:8000 -w yeti --callable webapp --processes 4 --threads 2 --stats 127.0.0.1:9191
Restart=always
KillSignal=SIGQUIT
Type=notify
StandardError=syslog
NotifyAccess=all

[Install]
WantedBy=multi-user.target
```

**Comment by sebdraven** [comment]
I think the problem is pip install -r requirements.txt --user . If you add --user, the different libs are installed in your home. the best way, it's to use a virtualenv

**Comment by tomchop** [comment]
Yeah, unless you're doing this with the `yeti` user (as specified in you uwsgi service definition), it will fail

**Comment by phutch11** [comment]
Yes all is performed with the Yeti user. Without the --user parameter the pip install procedures fails.

This server has been up for a number of years and contains a lot of data. It's sole purpose is to Yeti and doesn't perform any other function. It undergoes regular updates via the usual means with little to no trouble.

This is the first time that this message appears after an upgrade. I am sure something is amiss - I just cant put my finger on it.

**Comment by tomchop** [comment]
It will be easier to debug if you hop on the [slack channel](https://join.slack.com/t/yeti-platform/shared_invite/zt-864uilr4-zK7RCN~2VvoEnXV4RgAgAw)

**Comment by phutch11** [comment]
Closing issue.

Problem was that uwsgi was depending on Python2.7 - removing this dependency was messy.

This was not a Yeti issue but a uwsgi/python one.

Rebuilt server using modified bootstrap replacing all python related entries to suit Python3.

New server is up and running without issues using full Python3 suite.

All data restored.

PS: Thanks Tomchop for your time in [slack channel](https://join.slack.com/t/yeti-platform/shared_invite/zt-864uilr4-zK7RCN~2VvoEnXV4RgAgAw)

**Comment by fl0wc0ntr0l** [comment]
@phutch11 Do you mind sharing how exactly you were able to remove the Python 2.7 dependency? I am encountering the same issue on a fresh CentOS 7 install after mistakenly running the 2.7-based bootstrap script.

**Comment by phutch11** [comment]
@fl0wc0ntr0l glad to hear you got it sorted.

In case anyone else comes here with the same question.

**TL;DR Version:**
- Took Mongo dump of data and backed up Storage folder (Contains your images)
- Built a fresh VM and removed python2.7
- Followed most of the build steps manually, replacing all instances of python2.7 to python3 equivalents under "dependencies"
- After cloning repo, use PIP3 for your "requirements" installation (including uwsgi-plugin-python3)
- Rest of install is fine
- Restore Mongo backup and copy Storage folder contents across

The bootstrap needed modifications so I had to install everything manually. Made notes along the way and ended up with the following. This worked for me and suits my production environment containing my research and dossiers of around 7 years for almost 400 threat groups. It wont be for everyone but will help answer questions.

**Full Steps:**
```
# Preparation (Post OS Cleanup)
sudo apt install curl

# Remove Python2
sudo apt purge python2.7-minimal

sudo apt upgrade

curl -sS https://dl.yarnpkg.com/debian/pubkey.gpg | sudo apt-key add -
echo "deb https://dl.yarnpkg.com/debian/ stable main" | sudo tee /etc/apt/sources.list.d/yarn.list

# Mongo
wget -qO - https://www.mongodb.org/static/pgp/server-4.2.asc | sudo apt-key add -
echo "deb [ arch=amd64,arm64 ] https://repo.mongodb.org/apt/ubuntu bionic/mongodb-org/4.2 multiverse" | sudo tee /etc/apt/sources.list.d/mongodb-org-4.2.list
sudo apt-get update
sudo apt-get install -y mongodb-org

# Dependencies
sudo apt-get install build-essential git python3-dev mongodb-org redis-server libcurl4 libxml2-dev libxslt-dev zlib1g-dev python3-virtualenv python3-pip nginx yarn -y

# Clone Repo
cd /opt
sudo git clone https://github.com/yeti-platform/yeti.git

### Perform the following manually or run modified - sudo sh ./ubuntu_bootstrap.sh from Documents folder (no -H or uwsgi fails)

# Requirements
cd yeti
pip3 install -r requirements.txt
pip3 install uwsgi
sudo apt-get install uwsgi-plugin-python3
sudo yarn install

# Configure services
useradd yeti
sudo mkdir /var/log/yeti
sudo chown yeti /var/log/yeti
sudo cp extras/systemd/*.service /etc/systemd/system/
sudo systemctl enable mongod.service
sudo systemctl enable yeti_uwsgi.service
sudo systemctl enable yeti_oneshot.service
sudo systemctl enable yeti_feeds.service
sudo systemctl enable yeti_exports.service
sudo systemctl enable yeti_analytics.service
sudo systemctl enable yeti_beat.service
sudo systemctl daemon-reload
sudo chown -R yeti:yeti /opt/yeti
sudo chmod +x /opt/yeti/yeti.py

# Configure nginx
sudo rm /etc/nginx/sites-enabled/default
sudo cp extras/nginx/yeti /etc/nginx/sites-available/
sudo ln -s /etc/nginx/sites-available/yeti /etc/nginx/sites-enabled/yeti
sudo service nginx restart

# Start services
echo "[+] Starting services..."
sudo systemctl start mongod.service
sudo systemctl start yeti_oneshot.service
sleep 5
sudo systemctl start yeti_feeds.service
sudo systemctl start yeti_exports.service
sudo systemctl start yeti_analytics.service
sudo systemctl start yeti_beat.service
sudo systemctl start yeti_uwsgi.service
echo "[+] Yeti succesfully installed. Webserver listening on tcp/80"

* Copy Storage folder from old location
* Re-add account API information
* Sync old and new yeti installs and dbsyncs
* Restore Mongo backup
```
